### PR TITLE
Changing syntax to use MATERIALIZED VIEW for continuous aggregates

### LINKED ISSUE #2233: Use materialized view for continuous aggregates
Continuous aggregates are created using `CREATE VIEW` but this lacks support for adding table access methods and tablespaces as well as other storage options. Since continuous aggregates are backed by a materialization table, a more natural fit is to use `CREATE MATERIALIZED VIEW`.

This involves making sure that:

- [x] `CREATE MATERIALIZED VIEW` can be used to create a continuous aggregate.
- [x] `ALTER MATERIALIZED VIEW` can be used to alter a continuous aggregate.
- [x] `DROP MATERIALIZED VIEW` can be used to drop a continuous aggregate.
- [x] `DROP VIEW` on a continuous aggregate fails even though it is a view object.
- [x] `ALTER VIEW` on a continuous aggregate fails even though it is a view object.

### PR DESCRIPTION
**Pull request by mkindahl**
We change the syntax for defining continuous aggregates to use `CREATE
MATERIALIZED VIEW` rather than `CREATE VIEW`. The command still creates
a view, while `CREATE MATERIALIZED VIEW` creates a table.  Raise an
error if `CREATE VIEW` is used to create a continuous aggregate and
redirect to `CREATE MATERIALIZED VIEW`.

In a similar vein, `DROP MATERIALIZED VIEW` is used for continuous
aggregates and continuous aggregates cannot be dropped with `DROP
VIEW`.

Continuous aggregates are altered using `ALTER MATERIALIZED VIEW`
rather than `ALTER VIEW`, so we ensure that it works for `ALTER
MATERIALIZED VIEW` and gives an error if you try to use `ALTER VIEW` to
change a continuous aggregate.

Note that we allow `ALTER VIEW ... SET SCHEMA` to be used with the
partial view as well as with the direct view, so this is handled as a
special case.

Fixes #2233

### CODE REVIEW

File: src/process_utility.c [hunk 1]
```diff
@@ -3124,35 +3164,44 @@ process_altertable_reset_options(AlterTableCmd *cmd, Hypertable *ht)
 }
 
 static DDLResult
-process_viewstmt(ProcessUtilityArgs *args)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/process_utility.c:3127]: Added check for this.

File: tsl/src/continuous_aggs/create.c [hunk 2]
```diff
@@ -1751,44 +1749,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
+													  query_string,
+													  pstmt,
+													  with_clause_options);
+			break;
+		default:
+			elog(ERROR, "invalid statement for creating continuous aggregate");
+			break;
+	}
+
+	return false;
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply tsl/src/continuous_aggs/create.c]: Fixed.

File: src/process_utility.c [hunk 3]
```diff
@@ -2756,6 +2751,39 @@ process_altertable_start_view(ProcessUtilityArgs *args)
...
+				 errhint("Use ALTER MATERIALIZED VIEW to alter a continuous aggregate.")));
+
+	/* Check if this is an internal view of a continuous aggregate and give
+	 * error if attempts are made to alter them. */
+	namestrcpy(&view_name, get_rel_name(relid));
+	namestrcpy(&view_schema, get_namespace_name(get_rel_namespace(relid)));
+	cagg = ts_continuous_agg_find_by_view_name(NameStr(view_schema), NameStr(view_name));
+	vtyp = ts_continuous_agg_view_type(&cagg->data, NameStr(view_schema), NameStr(view_name));
+
+	if (vtyp == ContinuousAggPartialView || vtyp == ContinuousAggDirectView)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/process_utility.c:2779]: I'll clarify the situation when squashing them.

File: test/sql/updates/setup.continuous_aggs.v2.sql [hunk 4]
```diff
@@ -76,7 +76,7 @@ REFRESH MATERIALIZED VIEW mat_before;
 -- but realtime agg view definition is not stable across versions
 CREATE SCHEMA cagg;
 
-CREATE VIEW cagg.realtime_mat
+CREATE MATERIALIZED VIEW cagg.realtime_mat
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply test/sql/updates/setup.continuous_aggs.v2.sql]: I removed this change for now and will check what the update tests say.

File: src/process_utility.c [hunk 5]
```diff
@@ -3131,33 +3171,59 @@ process_altertable_reset_options(AlterTableCmd *cmd, Hypertable *ht)
...
+	switch (stmt->relkind)
 	{
-		parse_results = ts_continuous_agg_with_clause_parse(cagg_options);
-		is_cagg = DatumGetBool(parse_results[ContinuousEnabled].parsed);
-	}
+		case OBJECT_MATVIEW:
 
-	if (!is_cagg)
-		return DDL_CONTINUE;
+			/* is this a continuous agg */
```
Comment (erimatnor) [inline src/process_utility.c:3200]:
```suggestion
			/* Check for creation of continuous aggregate */
```

File: src/process_utility.c [hunk 6]
```diff
@@ -3131,33 +3171,59 @@ process_altertable_reset_options(AlterTableCmd *cmd, Hypertable *ht)
...
-				 errmsg("only timescaledb parameters allowed in WITH clause for continuous "
-						"aggregate")));
+			if (cagg_options)
+			{
+				parse_results = ts_continuous_agg_with_clause_parse(cagg_options);
+				is_cagg = DatumGetBool(parse_results[ContinuousEnabled].parsed);
+			}
+
+			if (!is_cagg)
+				return false;
```
Comment (erimatnor) [inline src/process_utility.c:3210]: wrong return type:

```suggestion
				return DDL_CONTINUE;
```

File: src/process_utility.c [hunk 7]
```diff
@@ -3131,33 +3171,59 @@ process_altertable_reset_options(AlterTableCmd *cmd, Hypertable *ht)
...
+
+			if (pg_options != NIL)
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("only timescaledb parameters allowed in WITH clause for continuous "
+								"aggregate")));
```
Comment (erimatnor) [inline src/process_utility.c:3216]:
```suggestion
						 errmsg("unsupported combination of storage parameters"),
						 errdetail("A continuous aggregate does not support standard storage parameters."),
						 errhint("Use only parameters with the \"timescaledb.\" prefix when creating a continuous aggregate."));
```

File: tsl/src/continuous_aggs/create.c [hunk 8]
```diff
@@ -1751,44 +1751,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
+
+	if (get_relname_relid(stmt->into->rel->relname, nspid))
 	{
 		ereport(ERROR,
 				(errcode(ERRCODE_DUPLICATE_TABLE),
-				 errmsg("continuous aggregate query \"%s\" already exists", stmt->view->relname),
+				 errmsg("continuous aggregate query \"%s\" already exists",
+						stmt->into->rel->relname),
 				 errhint("drop and recreate if needed.  This will drop the underlying "
 						 "materialization")));
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/create.c:1777]: Since you are changing this code anyway, we should also fix this hint since it is not following error guidelines. I actually don't think the hint adds any value (it is self-evident) so it can be removed. Otherwise I suggest:

```suggestion
				 errmsg("continuous aggregate \"%s\" already exists",
						stmt->into->rel->relname),
				 errhint("Drop or rename the existing continuous aggregate first or use another name.")));
```

File: tsl/src/continuous_aggs/create.c [hunk 9]
```diff
@@ -1751,44 +1751,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
+		.query = stmt->into->viewQuery,
+		.options = stmt->into->options,
+		.aliases = stmt->into->colNames,
+	};
 
-	nspid = RangeVarGetCreationNamespace(stmt->view);
-	if (get_relname_relid(stmt->view->relname, nspid))
+	nspid = RangeVarGetCreationNamespace(stmt->into->rel);
+
+	if (get_relname_relid(stmt->into->rel->relname, nspid))
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/create.c:1770]: Shouldn't this honor `IF NOT EXISTS`? (Looks like we are always failing here in case of an existing aggregate). Do we have a test for it?

File: tsl/src/continuous_aggs/create.c [hunk 10]
```diff
@@ -1751,44 +1751,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
+		case T_CreateTableAsStmt:
+			return continuous_agg_process_create_stmt(castNode(CreateTableAsStmt, stmt),
+													  query_string,
+													  pstmt,
+													  with_clause_options);
+			break;
+		default:
+			elog(ERROR, "invalid statement for creating continuous aggregate");
+			break;
+	}
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/create.c:1802]: Since only one case, I would simplify to:

```suggestion
	if (!IsA(stmt, CreateTableAsStmt))
		elog(ERROR, "invalid statement for creating continuous aggregate");
		
	continuous_agg_process_create_stmt(castNode(CreateTableAsStmt, stmt), query_string, pstmt, with_clause_options);
```

But I am wondering why we would even get here with another type so perhaps we should just `castNode` it to the right type and rely on the assertion to check it for us?

File: tsl/src/continuous_aggs/create.c [hunk 11]
```diff
@@ -1751,44 +1751,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
 }
 
+DDLResult
+tsl_process_continuous_agg_viewstmt(Node *stmt, const char *query_string, void *pstmt,
+									WithClauseResult *with_clause_options)
+{
+	switch (nodeTag(stmt))
+	{
+		case T_CreateTableAsStmt:
+			return continuous_agg_process_create_stmt(castNode(CreateTableAsStmt, stmt),
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/create.c:1794]: The function `continuous_agg_process_create_stmt` has a different return type (bool) than this function (DDLResult).

File: tsl/src/continuous_aggs/create.c [hunk 12]
```diff
@@ -1751,44 +1751,59 @@ cagg_create(ViewStmt *stmt, Query *panquery, CAggTimebucketInfo *origquery_ht,
...
 }
 
-/* entry point for creating continuous aggregate view
- * step 1 : validate query
- * step 2: create underlying tables and views
- */
-DDLResult
-tsl_process_continuous_agg_viewstmt(ViewStmt *stmt, const char *query_string, void *pstmt,
-									WithClauseResult *with_clause_options)
+static bool
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/create.c:1754]: Mismatching type of return value: bool here, but returns DDL_DONE. I would probably make this return void since there is only one possible return value.

**Review by erimatnor — APPROVED** [review]
Approving, but there are a number of nits, issues to fix. Most pressing are several places where the return type and return values do not match.

I also think there's a test case missing for `CREATE MATERIALIZED VIEW IF NOT EXISTS`

File: tsl/src/continuous_aggs/create.c (hunk 9 above)
Comment (mkindahl) [reply tsl/src/continuous_aggs/create.c:1771]: Added check for it working in a similar way to how it works for normal tables. And added test case.

File: tsl/src/continuous_aggs/create.c (hunk 11 above)
Comment (mkindahl) [reply tsl/src/continuous_aggs/create.c]: These are merged.

File: src/process_utility.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/process_utility.c:2784]: Rewrote it to mention that it is `ALTER VIEW ... SET SCHEMA` that we handle.

File: tsl/test/isolation/expected/continuous_aggs_multi.out [hunk 13]
```diff
@@ -94,7 +94,8 @@ step Setup2:
     END; $$ LANGUAGE plpgsql;
 
 step AlterLag1: alter view continuous_view_1 set (timescaledb.refresh_lag = 10);
-R1: LOG:  materializing continuous aggregate public.continuous_view_1: nothing to invalidate, new range up to 15
+ERROR:  cannot alter continuous aggregate using ALTER VIEW
```
Comment (gayyappan) [inline tsl/test/isolation/expected/continuous_aggs_multi.out:97]: This error needs to be fixed. It alters the output for the .out file.

File: tsl/test/isolation/expected/continuous_aggs_multi.out [hunk 14]
```diff
@@ -119,11 +120,12 @@ step Refresh2: REFRESH MATERIALIZED VIEW continuous_view_2; <waiting ...>
 step UnlockCompleted: ROLLBACK;
 step Refresh2: <... completed>
 step UnlockMat1: ROLLBACK;
-R1: LOG:  materializing continuous aggregate public.continuous_view_1: nothing to invalidate, new range up to 30
+R1: LOG:  materializing continuous aggregate public.continuous_view_1: nothing to invalidate, new range up to 45
```
Comment (gayyappan) [inline tsl/test/isolation/expected/continuous_aggs_multi.out:123]: Is this diff also related to the alter view fialing?

File: tsl/test/isolation/expected/continuous_aggs_multi.out (hunk 13 above)
Comment (mkindahl) [reply tsl/test/isolation/expected/continuous_aggs_multi.out]: Fixed.

File: tsl/test/isolation/expected/continuous_aggs_multi.out (hunk 14 above)
Comment (mkindahl) [reply tsl/test/isolation/expected/continuous_aggs_multi.out]: Yup. Fixed.
